### PR TITLE
[Test] Add conv2d Integrated test

### LINKED ISSUE #705: Conv2d backward fix
- [x] 1 channel 1 filter size
- below were bug inside recorder.py
  - [x] multi channel 1 filter -> weight grad is transposed in wrong order
  - [x] 1 channel multi filter -> weight grad is transposed in wrong order
- [x] 1 channel multi filter -> multiply 180 rotated weight

### PR DESCRIPTION
**Pull request by zhoonit**
- ~**#678** [Test/Util] Update ini / debug info~
- ~**#678** [IntegratedTest] Add batch normalization test~
- ~**#703** [Test/Py] Add translayer~
- [Conv] Fix convlayer backwarding
```
There was a miscalculation doing conv2d backwarding.

This patch fixes the issue.

With this patch, some test fails. I'll take a look and fix accordingly
within this commit

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: Jihoon Lee <jhoon.it.lee@samsung.com>
```
- [Test] Add conv2d Integrated test
```
dfc268a
This patch add conv2d integrated test

Model test passes, layer test failes due to bug fix

**Changes proposed in this PR:**
- add conv2d integrated test/generator
- add transLayer for channelfirst <-> channel last conversion

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: Jihoon Lee <jhoon.it.lee@samsung.com>
```

resolves #705, #544

### CODE REVIEW

File: nntrainer/src/conv2d_layer.cpp [hunk 1]
```diff
@@ -315,20 +312,32 @@ sharedConstTensors Conv2DLayer::backwarding(sharedConstTensors derivatives,
    *       x [(input_dim.height() + padding[0]*2)
    *           *(input_dim.width() + padding[1]*2) (width)]
    */
+  using uint = unsigned int;
+
+  Tensor ret(in_dim.batch(), in_dim.channel(), in_dim.height() + padding[0] * 2,
+             in_dim.width() + padding[1] * 2);
+  ret.setZero();
```
Comment (kparichay) [inline nntrainer/src/conv2d_layer.cpp:319]: most likely out of scope for this PR
but is this serZero needed? ret is used with conv2d_gemm and gemm does not need zeroing out output if beta (multiplier for output) is set to zero which is done at 545.
If removing this line changes output, then maybe another bug exist?

Comment (zhoonit) [reply nntrainer/src/conv2d_layer.cpp:319]: I think it is not needed. I'll remove this. If an error exists I'll leave an issue instead.

File: nntrainer/src/conv2d_layer.cpp [hunk 2]
```diff
@@ -315,20 +312,32 @@ sharedConstTensors Conv2DLayer::backwarding(sharedConstTensors derivatives,
...
 
   TensorDim kdim(ret.channel(), filter_size, kernel_size[0], kernel_size[1]);
 
-  std::vector<float> imkernel(kdim.getDataLen());
+  uint kernel_total_size = kernel_size[0] * kernel_size[1];
 
-  unsigned int count = 0;
-  float *d = imkernel.data();
+  Tensor imKernel(1, 1, in_dim.channel(), filter_size * kernel_total_size);
+  float *d = imKernel.getData();
```
Comment (kparichay) [inline nntrainer/src/conv2d_layer.cpp:326]: rename to imKernel_data or something more verbose?

File: test/unittest/unittest_nntrainer_models.cpp [hunk 3]
```diff
@@ -461,41 +472,75 @@ using INI = IniTestWrapper;
...
  */
+// clang-format off
 INI fc_sigmoid_mse(
   "fc_sigmoid_mse",
-  {nn_base + "learning_rate=1 | optimizer=sgd | loss=mse | batch_size = 3",
-   I("input") + input_base + "input_shape = 1:1:3",
-   I("dense") + fc_base + "unit = 5", I("act") + sigmoid,
-   I("dense_1") + fc_base + "unit = 10", I("act_1") + softmax});
+  {
+    nn_base + "learning_rate=1 | optimizer=sgd | loss=mse | batch_size = 3",
```
Comment (kparichay) [inline test/unittest/unittest_nntrainer_models.cpp:479]: minor suggestion, randomize batch size in the test cases below where batch size `== 1` and `> 1` both happens?

File: nntrainer/src/conv2d_layer.cpp (hunk 1 above)
Comment (zhoonit) [reply nntrainer/src/conv2d_layer.cpp:319]: Erased

File: test/unittest/unittest_nntrainer_models.cpp (hunk 3 above)
Comment (zhoonit) [reply test/unittest/unittest_nntrainer_models.cpp:479]: Added a new tc with one batch

File: nntrainer/src/conv2d_layer.cpp (hunk 2 above)
Comment (zhoonit) [reply nntrainer/src/conv2d_layer.cpp]: changed to `imKernel_raw`

**Review by jijoongmoon — APPROVED** [review]
LGTM. Once this is merged to main, I recommend to run MNIST Test again.

**Comment by kparichay** [comment]
> LGTM. Once this is merged to main, I recommend to run MNIST Test again.

@zhoonit Please run the MNIST test once and ensure that it reaches the ballpark of desired accuracy. Matching with tf is good but can be done after merging this PR too.

**Comment by zhoonit** [comment]
> > LGTM. Once this is merged to main, I recommend to run MNIST Test again.
> 
> @zhoonit Please run the MNIST test once and ensure that it reaches the ballpark of desired accuracy. Matching with tf is good but can be done after merging this PR too.

Do you have the number or how to generate a reference? Should I just match with tf result in confluence accuracy?

**Comment by kparichay** [comment]
You can match with tf accuracy in confluence, or just train with tf locally and match.

**Comment by zhoonit** [comment]
reference:            #1500/1500 - Training Loss:   0.042008 >> [ Accuracy:  99.596774  - Valiadtion Loss :   0.04159 ]
previous (09/25): #1500/1500 - Training Loss: 0.0254604 >> [ Accuracy: 99.7984% - Validation Loss : 0.0251934 ] 
this patch:           #1500/1500 - Training Loss: 0.0183812 >> [ Accuracy: 99.8992% - Validation Loss : 0.0181687 ]

**Review by kparichay — APPROVED** [review]
LGTM
